src/vaults/restaking/AffineReStaking.sol

```diff
@@ -0,0 +1,122 @@
+    // Token approval
+    bytes32 public constant APPROVED_TOKEN = keccak256("APPROVED_TOKEN");
+
+    IWETH public WETH;
+    // token amount from user
+    mapping(address => mapping(address => uint256)) public balance;
+
+    // event id for each event
+    uint256 private eventId;
+
+    // paused deposit
+    uint256 public depositPaused;
+
+    modifier whenDepositNotPaused() {
+        require(depositPaused == 0, "AR: deposit paused");
+        _;
+    }
+
+    function pauseDeposit() external onlyGovernance {
+        depositPaused = 1;
+    }
+
+    function resumeDeposit() external onlyGovernance {
+        depositPaused = 0;
+    }
+
+    // Approve token for deposit
+    function approveToken(address _token) external onlyGovernance {
+        if (hasRole(APPROVED_TOKEN, _token)) revert ReStakingErrors.AlreadyApprovedToken();
+        _grantRole(APPROVED_TOKEN, _token);
+    }
+
+    // Revoke token
+    function revokeToken(address _token) external onlyGovernance {
+        if (!hasRole(APPROVED_TOKEN, _token)) revert ReStakingErrors.NotApprovedToken();
+        if (ERC20(_token).balanceOf(address(this)) > 0) revert ReStakingErrors.NonZeroTokenBalance();
+
+        _revokeRole(APPROVED_TOKEN, _token);
+    }
+
+    event Deposit(uint256 indexed eventId, address indexed depositor, address indexed token, uint256 amount);
+    // deposit token for
+
+    function depositFor(address _token, address _for, uint256 _amount) external whenNotPaused whenDepositNotPaused {
+        if (_amount == 0) revert ReStakingErrors.DepositAmountCannotBeZero();
+        if (_for == address(0)) revert ReStakingErrors.CannotDepositForZeroAddress();
+        if (!hasRole(APPROVED_TOKEN, _token)) revert ReStakingErrors.TokenNotAllowedForStaking();
+
+        balance[_token][_for] += _amount;
+
+        emit Deposit(++eventId, _for, _token, _amount);
+
+        ERC20(_token).safeTransferFrom(msg.sender, address(this), _amount);
+    }
+
+    function depositETHFor(address _for) external payable whenNotPaused whenDepositNotPaused {
+        if (msg.value == 0) revert ReStakingErrors.DepositAmountCannotBeZero();
+        if (_for == address(0)) revert ReStakingErrors.CannotDepositForZeroAddress();
+        if (!hasRole(APPROVED_TOKEN, address(WETH))) revert ReStakingErrors.TokenNotAllowedForStaking();
+
+        balance[address(WETH)][_for] += msg.value;
+        emit Deposit(++eventId, _for, address(WETH), msg.value);
+
+        WETH.deposit{value: msg.value}();
+    }
+
+    event Withdraw(uint256 indexed eventId, address indexed withdrawer, address indexed token, uint256 amount);
+
+    function withdraw(address _token, uint256 _amount) external whenNotPaused {
+        if (_amount == 0) revert ReStakingErrors.WithdrawAmountCannotBeZero();
+        if (balance[_token][msg.sender] < _amount) revert ReStakingErrors.InvalidWithdrawalAmount();
+
+        balance[_token][msg.sender] -= _amount;
+        emit Withdraw(++eventId, msg.sender, _token, _amount);
+
```

Comment:
should prob unwrap the eth if token is weth and send it back as `value` since that's what they deposited